README.md
# 0. What is TerarkDB
TerarkDB is a RocksDB replacement with optimized tail latency, throughput and compression etc. In most cases you can migirate your existing RocksDB instance to TerarkDB without any
drawbacks.

- [All-in-one Docs](https://bytedance.feishu.cn/docs/doccnZmYFqHBm06BbvYgjsHHcKc#)
- [Slack Channel](https://join.slack.com/t/terarkdb/shared_invite/zt-k67ibxwn-PMn~x4yHqV0YD30FqNjqsQ)

**NOTES**
- TerarkDB was only tested and production ready under Linux platform
- Language bindings except C/C++ are not fully tested yet.
- Existing data can be migirated from RocksDB directly to TerarkDB, but cannot migrate back to RocksDB.
- TerarkDB was forked from RocksDB v5.18.3.


## Performance Overview
- RocksDB v6.12
- Server
  - Intel(R) Xeon(R) Gold 5218 CPU @ 2.30GHz (2 Sockets, 32 cores 64 threads)
  - 376 GB DRAM
  - NVMe TLC SSD (3.5 TB)
- Bench Tools & Workloads
  - use `db_bench`
  - 10 client threads, 20GB requests per thread
  - key = 24 bytes, value = 2000 bytes
  - `heavy_write` means 90% write operations
  - `heavy_read` means 90% read operations


![](docs/images/compare_rocksdb.png)


# 1. Use TerarkDB

## Prerequisite
If you enabled TerarkZipTable support (`-DWITH_TERARK_ZIP=ON`), you should install `libaio` before compile TerarkDB:

`sudo apt-get install libaio-dev`

If this is your first time using TerarkDB, we recommend you to use without TerarkZipTable by changing `-DWITH_TERARK_ZIP` to `OFF` in `build.sh`.

## Method 1: Use CMake subdirectory (Recommend)

1) Clone

```
cd {YOUR_PROJECT_DIR}
git submodule add https://github.com/bytedance/terarkdb.git

cd terarkdb && git submodule update --init --recursive
```

2) Edit your Top Project's CMakeLists.txt

```
add_subdirectory(terarkdb)
target_link_libraries({YOUR_TARGET} terarkdb)
```

3) Important Default Options

- CMAKE_BUILD_TYPE: RelWithDebInfo
- WITH_JEMALLOC: ON
  - Use Jemalloc or Not (If you are using a different malloc library, change to OFF)
- WITH_TESTS: OFF
  - Build test cases
- WITH_TOOLS: OFF
  - Build with TerarkDB tools (e.g. db_bench, ldb etc)
- WITH_TERARK_ZIP: OFF
  - Build with TerarkZipTable


### Notes
- TerarkDB is built with zstd, lz4, snappy, zlib, gtest, boost by default, if you need these libraries, you can remove them from your higher level application.


## Method 2: Link as static library

1) clone & build

```
git clone https://github.com/bytedance/terarkdb.git

cd terarkdb && git submodule update --init --recursive

./build.sh
```

2) linking

Directory:

```
  terarkdb/
        \___ output/
                \_____ include/
                \_____ lib/
                         \___ libterarkdb.a
                         \___ libzstd.a
                         \___ ...
```

We didn't archieve all static libraries together yet, so you have to pack all libraries to your target:

```
-Wl,-Bstatic \
-lterarkdb -lbz2 -ljemalloc -llz4 -lsnappy -lz -lzstd \
-Wl,-Bdynamic -pthread -lgomp -lrt -ldl -laio
```


# 2. Usage
## 2.1. BlockBasedTable
```c++
#include <cassert>
#include "rocksdb/db.h"

TERARKDB_NAMESPACE::DB* db;
TERARKDB_NAMESPACE::Options options;

// Your options here
options.create_if_missing = true;
options.wal_bytes_per_sync = 32768;
options.bytes_per_sync = 32768;

// Open DB
auto status = TERARKDB_NAMESPACE::DB::Open(options, "/tmp/testdb", &db);

// Operations
std::string value;
auto s = db->Put(TERARKDB_NAMESPACE::WriteOptions(), "key1", "value1");
s = db->Get(TERARKDB_NAMESPACE::ReadOptions(), "key1", &value);
assert(s.ok());
assert("value1" == value);

s = db->Delete(TERARKDB_NAMESPACE::WriteOptions(), "key1");
assert(s.ok());
```

Or manually set table format and table options:

```c++
#include <cassert>
#include "rocksdb/db.h"
#include "rocksdb/options.h"
#include "rocksdb/table.h"

TERARKDB_NAMESPACE::DB* db;
TERARKDB_NAMESPACE::Options options;

// Your db options here
options.create_if_missing = true;
options.wal_bytes_per_sync = 32768;
options.bytes_per_sync = 32768;

// Manually specify target table and table options
TERARKDB_NAMESPACE::BlockBasedTableOptions table_options;
table_options.block_cache = TERARKDB_NAMESPACE::NewLRUCache(32ULL << 30, 8, false);
table_options.block_size = 8ULL << 10;
options.table_factory = std::shared_ptr<TERARKDB_NAMESPACE::TableFactory>
                          (NewBlockBasedTableFactory(table_options));

// Open DB
auto status = TERARKDB_NAMESPACE::DB::Open(options, "/tmp/testdb2", &db);

// Operations
std::string value;
auto s = db->Put(TERARKDB_NAMESPACE::WriteOptions(), "key1", "value1");
s = db->Get(TERARKDB_NAMESPACE::ReadOptions(), "key1", &value);
assert(s.ok());
assert("value1" == value);

s = db->Delete(TERARKDB_NAMESPACE::WriteOptions(), "key1");
assert(s.ok());
```

## 2.2. TerarkZipTable
```c++
#include <cassert>
#include "rocksdb/db.h"
#include "rocksdb/options.h"
#include "rocksdb/table.h"
#include "table/terark_zip_table.h"

TERARKDB_NAMESPACE::DB* db;
TERARKDB_NAMESPACE::Options options;

// Your db options here
options.create_if_missing = true;
options.wal_bytes_per_sync = 32768;
options.bytes_per_sync = 32768;

// TerarkZipTable need a `fallback` options because you can indicate which LSM level you want to start using TerarkZipTable
// For example, by setting tzt_options.terarkZipMinLevel = 2, TerarkDB will use your fallback Table on level 0 and 1.
std::shared_ptr<TERARKDB_NAMESPACE::TableFactory> table_factory;
TERARKDB_NAMESPACE::BlockBasedTableOptions blockbased_options;
blockbased_options.block_size = 8ULL << 10;
table_factory.reset(NewBlockBasedTableFactory(blockbased_options));

TERARKDB_NAMESPACE::TerarkZipTableOptions tzt_options;
// TerarkZipTable requires a temp directory other than data directory, a slow device is acceptable
tzt_options.localTempDir = "/tmp";
tzt_options.indexNestLevel = 3;
tzt_options.sampleRatio = 0.01;
tzt_options.terarkZipMinLevel = 2; // Start using TerarkZipTable from level 2

table_factory.reset(TERARKDB_NAMESPACE::NewTerarkZipTableFactory(tzt_options, table_factory));

options.table_factory = table_factory;

// Open DB
auto status = TERARKDB_NAMESPACE::DB::Open(options, "/tmp/testdb2", &db);

// Operations
std::string value;
auto s = db->Put(TERARKDB_NAMESPACE::WriteOptions(), "key1", "value1");
s = db->Get(TERARKDB_NAMESPACE::ReadOptions(), "key1", &value);
assert(s.ok());
assert("value1" == value);

s = db->Delete(TERARKDB_NAMESPACE::WriteOptions(), "key1");
assert(s.ok());
```


# 3. Real-world Performance Improvement
TerarkDB has been deployed in lots of applications in Bytedance, in most cases TerarkDB can help to reduce latency spike and improve throughput tremendously.

### Disk Write
![](docs/images/disk_write.png)

### Get Latency (us)
![](docs/images/get_latency.png)


# 4. Contributing
- TerarkDB uses Github issues and pull requests to manage features and bug fixes.
- All PRs are welcome including code formating and refactoring.


# 5. License
- Apache 2.0

# 6. Users

Please let us know if you are using TerarkDB, thanks! (guokuankuan@bytedance.com)

- ByteDance (core online services)


HISTORY.md
# Rocksdb Change Log
<<<<<<< HEAD
=======
## 6.11 (6/12/2020)
### Bug Fixes
* Fix consistency checking error swallowing in some cases when options.force_consistency_checks = true.
* Fix possible false NotFound status from batched MultiGet using index type kHashSearch.
* Fix corruption caused by enabling delete triggered compaction (NewCompactOnDeletionCollectorFactory) in universal compaction mode, along with parallel compactions. The bug can result in two parallel compactions picking the same input files, resulting in the DB resurrecting older and deleted versions of some keys.
* Fix a use-after-free bug in best-efforts recovery. column_family_memtables_ needs to point to valid ColumnFamilySet.
* Let best-efforts recovery ignore corrupted files during table loading.
* Fix corrupt key read from ingested file when iterator direction switches from reverse to forward at a key that is a prefix of another key in the same file. It is only possible in files with a non-zero global seqno.
* Fix abnormally large estimate from GetApproximateSizes when a range starts near the end of one SST file and near the beginning of another. Now GetApproximateSizes consistently and fairly includes the size of SST metadata in addition to data blocks, attributing metadata proportionally among the data blocks based on their size.
* Fix potential file descriptor leakage in PosixEnv's IsDirectory() and NewRandomAccessFile().
* Fix false negative from the VerifyChecksum() API when there is a checksum mismatch in an index partition block in a BlockBasedTable format table file (index_type is kTwoLevelIndexSearch).
* Fix sst_dump to return non-zero exit code if the specified file is not a recognized SST file or fails requested checks.
* Fix incorrect results from batched MultiGet for duplicate keys, when the duplicate key matches the largest key of an SST file and the value type for the key in the file is a merge value.
* Fix "bad block type" error from persistent cache on Windows.

### Public API Change
* Flush(..., column_family) may return Status::ColumnFamilyDropped() instead of Status::InvalidArgument() if column_family is dropped while processing the flush request.
* BlobDB now explicitly disallows using the default column family's storage directories as blob directory.
* DeleteRange now returns `Status::InvalidArgument` if the range's end key comes before its start key according to the user comparator. Previously the behavior was undefined.
* ldb now uses options.force_consistency_checks = true by default and "--disable_consistency_checks" is added to disable it.
* DB::OpenForReadOnly no longer creates files or directories if the named DB does not exist, unless create_if_missing is set to true.
* The consistency checks that validate LSM state changes (table file additions/deletions during flushes and compactions) are now stricter, more efficient, and no longer optional, i.e. they are performed even if `force_consistency_checks` is `false`.
* Disable delete triggered compaction (NewCompactOnDeletionCollectorFactory) in universal compaction mode and num_levels = 1 in order to avoid a corruption bug.
* `pin_l0_filter_and_index_blocks_in_cache` no longer applies to L0 files larger than `1.5 * write_buffer_size` to give more predictable memory usage. Such L0 files may exist due to intra-L0 compaction, external file ingestion, or user dynamically changing `write_buffer_size` (note, however, that files that are already pinned will continue being pinned, even after such a dynamic change).
* In point-in-time wal recovery mode, fail database recovery in case of IOError while reading the WAL to avoid data loss.
* A new method `Env::LowerThreadPoolCPUPriority(Priority, CpuPriority)` is added to `Env` to be able to lower to a specific priority such as `CpuPriority::kIdle`.
* `DB::GetDbSessionId(std::string& session_id)` is added. `session_id` stores a unique identifier that gets reset every time the DB is opened. This DB session ID should be unique among all open DB instances on all hosts, and should be unique among re-openings of the same or other DBs. This identifier is recorded in the `LOG` file on the line starting with `DB Session ID:`.

### New Features
* sst_dump to add a new --readahead_size argument. Users can specify read size when scanning the data. Sst_dump also tries to prefetch tail part of the SST files so usually some number of I/Os are saved there too.
* Generate file checksum in SstFileWriter if Options.file_checksum_gen_factory is set. The checksum and checksum function name are stored in ExternalSstFileInfo after the sst file write is finished.
* Add a value_size_soft_limit in read options which limits the cumulative value size of keys read in batches in MultiGet. Once the cumulative value size of found keys exceeds read_options.value_size_soft_limit, all the remaining keys are returned with status Abort without further finding their values. By default the value_size_soft_limit is std::numeric_limits<uint64_t>::max().

bench.py
#!/usr/bin/python
#
# Copyright (c) 2020-present, Bytedance Inc.  All rights reserved.
# This source code is licensed under Apache 2.0 License.
#
# Usage:
#       ./bench.py $PWD/data 20 10 24,2000 readrandomwriterandom,10
#
#       ./bench.py $PWD/data 20 10 24,2000 readrandomwriterandom,90
#
import time
import io
import os
import subprocess
import sys
import requests
from datetime import datetime

# default values
KV_SIZE = "24, 500"
KSIZE = 24  # key size form KV_SIZE
VSIZE = 500 # value size from KV_SIZE
GB_PER_THREAD = 20
TOTAL_MEM_IN_GB = 64
THREADS = 16
DB_DIR = ""
BENCH_TYPE = "fillseq"
BENCH_ARGS = []

# collected result log
LOG_RESULT_FNAME = "log.txt"
# bench rocksdb output
LOG_BENCH_OUTPUT_FNAME = "output.txt"

BENCH_ENGINES = {'terarkdb':'./output/db_bench'}

def bench(records, key_size, value_size, engine, db_dir, exist_db):
    extra_flags = ''
    if engine == 'terarkdb':
        extra_flags = """
                         --use_terark_table=false
                         --blob_size=128
                      """

    if BENCH_TYPE == 'readrandomwriterandom':
        extra_flags += """
                         --readwritepercent=%s
                       """ % BENCH_ARGS[0]
    cmd = """
           {db_bench} \
           --benchmarks={bench_type}
	   --use_existing_db={exist_db}
           --sync=0
	   --db={db_dir}
	   --wal_dir={db_dir}
	   --bytes_per_sync=65536
           --wal_bytes_per_sync=65536
           --num={records}
           --threads={threads}
	   --num_levels=6
           --delayed_write_rate=209715200
	   --key_size={key_size}
	   --value_size={value_size}
	   --cache_numshardbits=6
	   --level_compaction_dynamic_level_bytes=true
	   --cache_index_and_filter_blocks=1
	   --pin_l0_filter_and_index_blocks_in_cache=0
	   --benchmark_write_rate_limit=0
	   --hard_rate_limit=3
	   --rate_limit_delay_max_milliseconds=1000000
	   --write_buffer_size=268435456
	   --max_write_buffer_number=6
	   --target_file_size_base=134217728
	   --max_bytes_for_level_base=536870912
	   --verify_checksum=1
	   --delete_obsolete_files_period_micros=62914560
	   --max_bytes_for_level_multiplier=10
	   --statistics=0
	   --stats_per_interval=1
	   --stats_interval_seconds=60
	   --histogram=1
	   --open_files=-1
	   --level0_file_num_compaction_trigger=4
	   --level0_slowdown_writes_trigger=1000
	   --level0_stop_writes_trigger=1000
           --num_high_pri_threads=3
           --num_low_pri_threads=10
           --mmap_read=true
           --compression_type=none
           --memtablerep=patricia_trie
           {extra_flags}
           """.format(records=records, 
                      key_size=key_size,
                      value_size=value_size, 
                      db_dir=db_dir, 
                      bench_type=BENCH_TYPE,
                      exist_db=exist_db,
                      threads=THREADS,
                      db_bench=BENCH_ENGINES[engine],
                      extra_flags=extra_flags)

    cmd = cmd.replace('\n',' ')
    log = open(LOG_BENCH_OUTPUT_FNAME, 'wb')
    log.write(cmd)
    log.flush()
    process = subprocess.Popen(cmd,
                               stdin=subprocess.PIPE,
                               stderr=log, 
                               stdout=log, 
                               shell=True)
    process.communicate()
    log.flush()
    log.close()
    print 'test finished: %s\n' % LOG_BENCH_OUTPUT_FNAME


def run(engine, db_dir):
    db_size_bytes = int(GB_PER_THREAD) * 1024 * 1024 * 1024
    records = db_size_bytes / (KSIZE + VSIZE)
    bench(records, KSIZE, VSIZE, engine, db_dir, 0)


def gather_result(engine):
    rst = {}
    for bench_type in [BENCH_TYPE]:
        rst[bench_type] = {}
        
        rst[bench_type] = {}

        with open(LOG_BENCH_OUTPUT_FNAME, 'rb') as f:
            lines = f.readlines()
            i = 0
            while i < len(lines):
                # get ops
                s = '%s' % bench_type
                if lines[i].find(s) == 0:
                    rst[bench_type]['ops'] = lines[i].split()[4]

                # get rest of them
                s = 'Microseconds per '
                if lines[i].find(s) >= 0:
                    ops_type = lines[i][17:-2]
                    rst[bench_type][ops_type] = {}
                    rst[bench_type][ops_type]['max'] = lines[i+2].split()[5]
                    rst[bench_type][ops_type]['percentiles'] = lines[i+3][13:-2]
                    i = i + 5
                else:
                    i = i + 1

    # print rst
    output = [('benchmark', 'kv bytes', 'ops', 'operation', 'max lat(us)', 'pct(us)')]
    for bench in rst:
            for t in ['read', 'write']:
                if rst[bench].has_key(t):
                    output.append( (bench, KV_SIZE, rst[bench]['ops'], t, rst[bench][t]['max'], rst[bench][t]['percentiles']) )

    with open(LOG_RESULT_FNAME, 'a') as f:
        for row in output:
            f.write('{0:<25} {1:<15} {2:<15} {3:<15} {4:<15} {5:<100}\n'.format(*row))

if __name__=='__main__':
    if not os.path.isfile(BENCH_ENGINES['terarkdb']):
        print 'db_bench not found, please check: %s', BENCH_ENGINES
        sys.exit()

    if len(sys.argv) != 6:
        print 'usage: ./bench.py [DB_DIR] [GB_PER_THREAD] [THREADS] [KV_SIZE] [BENCH_STR]\n'
        print '\t\tKV_SIZE: 24,500 means key size is 24, value size is 500'
        print '\t\tBENCH_STR: fillseq or readrandomwriterandom,90, the later one means 90% reads'
        sys.exit()

    DB_DIR = sys.argv[1]
    GB_PER_THREAD = sys.argv[2]
    THREADS = sys.argv[3]
    KV_SIZE = sys.argv[4]
    KSIZE, VSIZE = [int(i) for i in KV_SIZE.split(',')]

    BENCH_STR = sys.argv[5].split(",")
    BENCH_TYPE = BENCH_STR[0]
    BENCH_ARGS = BENCH_STR[1:]
    print 'bench_type = %s, bench_args = %s' % (BENCH_TYPE, BENCH_ARGS)

    LOG_RESULT_FNAME = 'rst_%s_gb_%s_thds.txt' % (GB_PER_THREAD, THREADS)

    for engine in ['terarkdb']:
        LOG_BENCH_OUTPUT_FNAME = "output_%s_%s_%s_%s.txt" % (engine, BENCH_TYPE, KSIZE, VSIZE)

        db_dir = '%s_%s' %  (DB_DIR, engine)
        print 'start engine : %s, db_dir = %s' % (engine, db_dir)

        with open(LOG_RESULT_FNAME, 'a') as f:
            f.write('[%s] GB_PER_THREAD: %s, THREADS = %s, KV_SIZE = %s, BENCH_STR = %s, DATA_DIR = %s\n' % (datetime.now().strftime("%Y-%m-%d %H:%M:%S"), GB_PER_THREAD, THREADS, KV_SIZE, BENCH_STR, db_dir) )

        run(engine, db_dir)
        gather_result(engine)

        with open(LOG_RESULT_FNAME, 'a') as f:
            f.write('\n\n')
